File: 28/solution.go
package _28

type KMP struct {
	pattern string
	dp      []int
}

func (k *KMP) Init(pattern string) {
	k.pattern = pattern
	m := len(pattern)
	k.dp = make([]int, m)
	j := 0
	for i := 1; i < m; i++ {
		for j > 0 && pattern[i] != pattern[j] {
			j = k.dp[j-1]
		}
		if pattern[i] == pattern[j] {
			j++
		}
		k.dp[i] = j
	}
}

func (k KMP) Search(txt string) int {
	j, m := 0, len(k.pattern)
	for i := 0; i < len(txt); i++ {
		for j > 0 && txt[i] != k.pattern[j] {
			j = k.dp[j-1]
		}
		if txt[i] == k.pattern[j] {
			j++
		}
		if j == m {
			return i - m + 1
		}
	}
	return -1
}

func strStr(haystack string, needle string) int {
	k := new(KMP)
	k.Init(needle)
	return k.Search(haystack)
}

File: 28/solution.py
class KMP:
    def __init__(self, pat: str):
        self.pat = pat
        m = len(pat)
        self.dp = [0] * m
        j = 0
        for i in range(1, m):
            while j > 0 and pat[i] != pat[j]:
                j = self.dp[j - 1]
            if pat[i] == pat[j]:
                j += 1
            self.dp[i] = j

    def search(self, txt: str):
        m = len(self.pat)
        n = len(txt)
        j = 0
        for i in range(n):
            while j > 0 and txt[i] != self.pat[j]:
                j = self.dp[j-1]
            if txt[i] == self.pat[j]:
                j += 1
            if j == m:
                return i - m + 1
        return -1

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        kmp = KMP(needle)
        return kmp.search(haystack)
